### ISSUE TITLE
println works selectively when using Run All

### ISSUE THREAD

**Issue by rvernica** [ibm-et/spark-kernel#129]
Hello,

I am using a recent revision of the code-base:
```
commit b8fe2c4f01c7916484c79c7043f646d5539ea05e
Merge: 376921c 5d1cd3a
Author: Chip Senkbeil <chip.senkbeil@gmail.com>
Date:   Mon Aug 24 16:08:20 2015 -0500
```
and Spark 1.4.1.

`println` seems to be working selectively when using `Run All`. I have a notebook with three cells. I load the notebook and the kernel starts OK. Then I click `Run All`. This is the output:
```
In [1]:
val a = sc.parallelize(List(1, 2, 3, 4))
println(a.count)

In [2]:
val b = a.reduce((x, y) => x * y)
println(a.count)
println(b)

4

In [3]:
val c = a.flatMap(x => List(x, x + 1))
println(a.count)
println(c.count)

4
8
```
Notice how there is no `println` output from the first cell (instead of one), only one `println` output from the second cell (instead of two), and correct `println` output from the third cell.

If I run the cells individually, it works fine:
```
In [4]:
val a = sc.parallelize(List(1, 2, 3, 4))
println(a.count)

4

In [5]:
val b = a.reduce((x, y) => x * y)
println(a.count)
println(b)

4
24

In [6]:
val c = a.flatMap(x => List(x, x + 1))
println(a.count)
println(c.count)

4
8
```

This might be related to #57

**Comment by rcsenkbeil** [comment]
There might be a race condition. Have you tried using `kernel.out.println` instead of `println` when running all of your cells? `kernel.out.println` directly sends a text-oriented message to the notebook versus `println`, which attempts to capture standard out and redirect it to the notebook.

**Comment by rvernica** [comment]
I tried using `kernel.out.println` and the behaviour is the same as `println`. Here is the output of `Run All`:
```
In [1]:
val a = sc.parallelize(List(1, 2, 3, 4))
kernel.out.println(a.count)

In [2]:
val b = a.reduce((x, y) => x * y)
kernel.out.println(a.count)
kernel.out.println(b)

4

In [3]:
val c = a.flatMap(x => List(x, x + 1))
kernel.out.println(a.count)
kernel.out.println(c.count)

4
8
```

**Comment by wellecks** [comment]
We're looking into this, and opened the above issue in Jupyter. The kernel's output stream introduces a slight delay between print statements, which causes the second print message in your example to arrive after the next cell has started execution. The kernel does send a `stream` message associated with the second `println`, but it does not get displayed by Jupyter.

For now, we'll wait for the response on the issue to decide whether to change things on our end.

**Comment by lbustelo** [comment]
@wellecks It seems that we got respond from the jupyter guys... Is it possible to move the `idle` message to later?

**Comment by wellecks** [comment]
@chipsenkbeil and I looked into this. We found two issues:

1. The `KernelOutputStream` flushes on a time interval; it was possible that an `idle` message was sent before the next flush interval. 
    - To fix, we added an explicit `flush` after code execution completes.

2. Messages moving from `KernelMessageRelay` to the outgoing `PubSocketActor` pass through several actors (e.g. `SignatureManager`, `SignatureProducer`), and the order of messages was not consistent throughout this process. As a result, an outgoing `stream` message that arrived at the `KernelMessageRelay` before an outgoing `idle` message may have actually been sent by the `PubSocketActor` after the `idle` message.
     - To fix, we added `OrderedSupport` to each actor encountered in the process of moving from `KernelMessageRelay` to `PubSocketActor`
